src/print.c

```diff
@@ -245,6 +259,183 @@ launch_preview (const char *filename,
   return retval;
 }
 
+static int
+pdf_get_actions_page (char        *filename,
+                      PDF_ACTIONS  act,
+                      int          page)
+{
+  GSubprocess *process;
+  GInputStream *stream;
+  GPtrArray *args = g_ptr_array_new_full(8, g_free);
+  char buffer[50] = { 0 };
+
+  g_ptr_array_add(args, LIBEXECDIR "/pdftoraw");
+  g_ptr_array_add(args, g_strconcat ("--file=", filename, NULL));
+  switch(act)
+  {
+    case PDF_TEST :
+       g_ptr_array_add(args, "--test");
+       break;
+    case PDF_NUM_PAGES :
+       g_ptr_array_add(args, "--pages");
+       break;
+    case PDF_WIDTH :
+       g_ptr_array_add(args, g_strdup_printf ("--width=%d", page));
+       break;
+    case PDF_HEIGHT :
+       g_ptr_array_add(args, g_strdup_printf ("--height=%d", page));
+       break;
+   default:
+       return 0;
+  }
+
+  process = g_subprocess_newv((const gchar * const *)args->pdata, G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe(process);
+    if (stream)
+    {
+      g_input_stream_read(stream, buffer, sizeof(buffer), NULL, NULL);
+      if (buffer[0])
+        return atoi(buffer);
+    }
+  }
+  return 0;
+}
+
+static unsigned char *
+pdf_get_data_page (char *filename,
+                   int   w,
+                   int   h,
+                   int   page)
+{
+  GSubprocess *process;
+  GInputStream *stream;
+  unsigned char *data = NULL;
+  GPtrArray *args = g_ptr_array_new_full(8, g_free);
+  int read = 0;
+  gsize total_read = 0;
+
+  g_ptr_array_add(args, LIBEXECDIR "/pdftoraw");
+  g_ptr_array_add(args, g_strconcat ("--file=", filename, NULL));
+  g_ptr_array_add(args, g_strdup_printf ("--raw=%d", page));
+  process = g_subprocess_newv((const gchar * const *)args->pdata, G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe(process);
+    if (stream)
+    {
+      data = (unsigned char *) g_malloc( w * h * 4);
+      while ((read = g_input_stream_read(stream, data + total_read, 1024, NULL, NULL)) > 0)
```

Comment:
You need to ensure that you don't read more data than `w * h * 4` here.

Reply:
Fixed by 9cb421f

Reply:
Fixed by 9cb421f